cql3/expr/prepare_expr.cc

```diff
@@ -1057,6 +1058,182 @@ ::shared_ptr<assignment_testable> as_assignment_testable(expression e) {
+            new_elements.reserve(tc.elements.size());
+
+            for (const expression& elem : tc.elements) {
+                new_elements.emplace_back(prepare_binop_lhs(elem, db, schema, printable_relation));
+            }
+
+            return tuple_constructor {
+                .elements = std::move(new_elements)
+            };
+        },
+        [&](const token& tk) -> expression {
+            std::vector<expression> prepared_token_args;
+            prepared_token_args.reserve(tk.args.size());
+
+            for (const expression& arg : tk.args) {
+                prepared_token_args.emplace_back(prepare_binop_lhs(arg, db, schema, printable_relation));
+            }
+
+            return token(std::move(prepared_token_args));
+        },
+        [&](const subscript& sub) -> expression {
+            const column_definition* sub_col = expr::visit(overloaded_functor {
+                [&](const unresolved_identifier& unin) -> const column_definition* {
+                    return resolve_column(unin, schema, printable_relation);
+                },
+                [](const column_value& cv) -> const column_definition* {
+                    return cv.col;
+                },
+                [](const auto& e) -> const column_definition* {
+                    on_internal_error(expr_logger,
+                        fmt::format("Only columns can be subscripted using the [] operator, got {}", e));
+                }
+            }, sub.val);
+
+            const abstract_type& sub_col_type = sub_col->column_specification->type->without_reversed();
+            if (!sub_col_type.is_map()) {
+                throw exceptions::invalid_request_exception(format("Column {} is not a map, cannot be used as a map", sub_col->name_as_text()));
+            }
+
+            return subscript {
+                .val = column_value(sub_col),
+                .sub = prepare_expression(sub.sub, db, schema.ks_name(), map_key_spec_of(*sub_col->column_specification))
+            };
+        },
+        [](const auto& e) -> expression {
+            on_internal_error(expr_logger, format("prepare_binop_lhs: Unhandled expression: {}", e));
+        }
+    }, lhs);
+}
+
+// Finds the type of a prepared LHS of binary_operator and creates a receiver with it.
+static lw_shared_ptr<column_specification> get_lhs_receiver(const expression& prepared_lhs, const schema& schema) {
+    return expr::visit(overloaded_functor{
+        [](const column_value& col_val) -> lw_shared_ptr<column_specification> {
+            return col_val.col->column_specification;
+        },
+        [](const subscript& col_val) -> lw_shared_ptr<column_specification> {
+            const column_value& sub_col = get_subscripted_column(col_val);
+            return map_value_spec_of(*sub_col.col->column_specification);
+        },
+        [&](const tuple_constructor& tup) -> lw_shared_ptr<column_specification> {
+            sstring tuple_name = "(";
```

Comment:
Please us std::ostringstream to avoid quadratic complexity.

Reply:
Done, I thought that strings work like vectors and allocate powers of 2, but they might not.

Reply:
sstring does not, it always allocates the minimum amount. It did hit other people too.